### PR TITLE
Integrate Arthas into Dev Mode

### PR DESCRIPTION
**Pull request by geoand**
[Arthas](https://arthas.aliyun.com/doc/en/index.html) is a Java diagnostic tool
open-sourced by Alibaba middleware team.
It is widely adopted and popular among the developers inside Alibaba.
Arthas helps developers in troubleshooting issues without modifying code or restarting servers.

This Pull Request essentially makes Arthas diagnostics available to a Quarkus application
running in Dev Mode without any user actions being necessary (other than opting into
using Arthas). 
When Arthas is started, its URL is printed in the console so users can easily access it.

Here is what the UI looks like after I issue the `help` command:

Screenshot from 2021-12-17 17-46-09

*Future work*

Currently this feature is opt-in via config that must be set when launching dev-mode. It might be interesting to enable this from a button in the Dev UI or an action in the console when the application is already running.

** UPDATE **

The PR now simply logs a message about the exact JBang script needed to launch and attach Arthas to the running Quarkus application.

### CODE REVIEW

**Comment by geoand** [comment]
@loicmathieu you might be interested in this 😉

File: core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * The versions of Arthas to use
+     */
+    @ConfigItem(defaultValue = "3.5.5")
+    public String version;
+
+    /**
+     * The directory into which to save the arthas-boot jar if it doesn't exist
+     */
+    @ConfigItem(defaultValue = "${user.home}/.quarkus")
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: @aloubyansky just wanted to check that you were not using this directory for other things that could interfere with this?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 2]
```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
...
+            r = () -> startArthas(toArthasJar(arthasDir), arthasHttpPort);
+        }
+        var arthasThread = new Thread(r);
+        arthasThread.setName("Arthas thread");
+        arthasThread.setDaemon(true);
+        arthasThread.start();
+    }
+
+    private Path downloadArthas(Path arthasDir, String version) {
+        var downloadURL = "https://arthas.aliyun.com/download/" + version;
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:274]: Can't say I'm excited with us downloading a jar by ourselves from a maybe unstable URL. So first, let's use a Maven Central URL to download this, the artifact is there AFAICS: https://search.maven.org/artifact/com.taobao.arthas/arthas-boot/3.5.5/jar .

Given how it works, I suppose using a dependency is out of the question?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 3]
```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
...
+            log.debug("Arthas downloaded");
+            return toArthasJar(arthasDir);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void startArthas(Path arthasJar, Integer httpPort) {
+        log.debug("Attempting to start Arthas");
+        ProcessBuilder processBuilder = new ProcessBuilder("java",
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:296]: I remember some more complicated logic to get the right java binary somewhere in the code base but maybe it's just a bad recollection.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 4]
```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
...
+            var arthasProcess = processBuilder.start();
+            try {
+                // attaching to the process takes a while
+                // TODO: we should probably tail the logs instead of waiting
+                Thread.sleep(5_000);
+            } catch (InterruptedException ignored) {
+
+            }
+            if (arthasProcess.isAlive()) {
+                log.infof("Arthas started and is accessible at http://127.0.0.1:%d", 8563);
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:321]: Maybe use localhost? I'm not sure the IPv4 IP works for someone configured with just IPv6.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added some comments and questions. The only thing I really want fixed is the download URL.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 4 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:321]: Just pushedthat change :)

File: core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java (hunk 1 above)
Comment (loicmathieu) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: Maybe use temp dir to have it automatically deleted at JVM exit ?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:274]: That's not enough unfortunately. For it to work, we need the entire zip.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 3 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:296]: Right, we do have that. Let me see if I can find it

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 2 above)
Comment (loicmathieu) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:275]: I think using an extension and a dependency should be feasible.
You can then locate the JAR from Maven dependency mechanism and launch it.

File: core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java (hunk 1 above)
Comment (gastaldi) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: The CLI uses the file `~/.quarkus/config.yaml` for adding new registries

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 5]
```diff
@@ -210,6 +224,117 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
...
+
+        }
+
+        try {
+            var arthasProcess = processBuilder.start();
+            try {
+                // attaching to the process takes a while
+                // TODO: we should probably tail the logs instead of waiting
+                Thread.sleep(5_000);
+            } catch (InterruptedException ignored) {
```
Comment (loicmathieu) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:320]: You should at least poll a few times each second or something like this.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 3 above)
Comment (gastaldi) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:296]: Is it in `io.quarkus.utilities.JavaBinFinder`?

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java]: Fixed

**Review by loicmathieu — CHANGES_REQUESTED** [review]
This is indead a good tool but I didn't use it since a long time !

I'm surprised it's comes OOTB and didn't come from an extension, this will looks very opinionated and make a strong dependency between the core and Arthas.

**Comment by loicmathieu** [comment]
I never use the web console, only the command line mode.
Maybe we can integrate the command line mode inside the quarkus dev console so there will not be complicated stuff to wait for the server to be ready. But maybe this would be more difficult.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:275]: Let me try it again

File: core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java (hunk 1 above)
Comment (aloubyansky) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: Right, let's not trash it. Perhaps it should have its own subdirectory at least.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:275]: Tried it again and it does not work - the entire zip is needed.

Comment (maxandersen) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:275]: https://github.com/alibaba/jbang-catalog/blob/main/jbang-catalog.json#L4

its in maven - its at https://repo1.maven.org/maven2/com/taobao/arthas/arthas-boot/3.5.5/arthas-boot-3.5.5-jar-with-dependencies.jar

and it exists as maven dependency too but you will need to resolve it without transitive dependencies.

File: core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: It does - it's added under `arthas`

Comment (maxandersen) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: definitely not a fan of us downloading it into ~/.quarkus - if we really want to do this I would resolve it as dependency.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: As I mentioned in another comment - the dependency is not enough.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:275]: OK, let me try this.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java]: I am not using this.

**Comment by maxandersen** [comment]
> I never use the web console, only the command line mode.
> Maybe we can integrate the command line mode inside the quarkus dev console so there will not be complicated stuff to wait for the server to be ready. But maybe this would be more difficult.

having it is devconsole is not really useful imo - then you can't see or interact with your own app.
if you want to use arthos in console mode you would just open another terminal and run it directly.

Thus I would much rather prefer it to just print out the necessary command ...` jbang arthas@alibaba 1234` or `java -jar arthas.jar 1234`

I'm not a fan of quarkus core devmode downloading tools like this...its not what users expect nor something we can easily productize.

File: core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:30]: Scratch that, see my comment below about the all-in-one jar.

**Comment by geoand** [comment]
I am not going to put any more effort into this honestly. If someone wants to bring it home, great.

Otherwise 🤷🏼

**Comment by geoand** [comment]
> I'm not a fan of quarkus core devmode downloading tools like this...its not what users expect nor something we can easily productize.

Why would we ever productize this?

And in any case, it's an opt-in feature...

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 5 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:319]: Sure, we can do this

**Comment by geoand** [comment]
> Thus I would much rather prefer it to just print out the necessary command ...` jbang arthas@alibaba 1234` or `java -jar arthas.jar 1234`

I pushed a second commit doing only this

**Comment by geoand** [comment]
Closing due to lack of interest

**Pull request closed without merging**